lib/credentials/chainable_temporary_credentials.js

```diff
@@ -0,0 +1,196 @@
+   * @option options masterCredentials [AWS.Credentials] (null) the master
+   *   (source) credentials used to get and refresh temporary credentials from
+   *   AWS STS. By default, {AWS.config.credentials} or
+   *   {AWS.config.credentialProvider} will be used.
+   * @option options tokenCodeFn [Function] (null) Function to provide
+   *   `TokenCode`, if `SerialNumber` is provided for profile in {params}. Function
+   *   is called with value of `SerialNumber` and `callback`, and should provide
+   *   the `TokenCode` or an error to the callback in the format
+   *   `callback(err, token)`.
+   * @example Creating a new credentials object for generic temporary credentials
+   *   AWS.config.credentials = new AWS.TemporaryCredentials();
+   * @example Creating a new credentials object for an IAM role
+   *   AWS.config.credentials = new AWS.TemporaryCredentials({
+   *     RoleArn: 'arn:aws:iam::1234567890:role/TemporaryCredentials',
+   *   });
+   * @see AWS.STS.assumeRole
+   * @see AWS.STS.getSessionToken
+   */
+  constructor: function ChainableTemporaryCredentials(options) {
+    AWS.Credentials.call(this);
+    options = options || {};
+    this.errorCode = 'ChainableTemporaryCredentialsProviderFailure';
+    this.expired = true;
+    this.refreshCallbacks = [];
+    this.tokenCodeFn = null;
+
+    var params = AWS.util.copy(options.params) || {}
+    if (params.RoleArn) {
+      params.RoleSessionName = params.RoleSessionName || 'temporary-credentials';
+    }
+    if (params.SerialNumber) {
+      if (!options.tokenCodeFn || (typeof options.tokenCodeFn !== 'function')) {
+        throw new AWS.util.error(
+          new Error('tokenCodeFn must be a function when params.SerialNumber is given'),
+          {code: this.errorCode}
+        );
+      } else {
+        this.tokenCodeFn = options.tokenCodeFn;
+      }
+    }
+    this.service = new STS({
+      params: params,
+      credentials: options.masterCredentials || AWS.config.credentials
+    });
+  },
+
+  /**
+   * Refreshes credentials using {AWS.STS.assumeRole} or
+   * {AWS.STS.getSessionToken}, depending on whether an IAM role ARN was passed
+   * to the credentials {constructor}.
+   *
+   * @callback callback function(err)
+   *   Called when the STS service responds (or fails). When
+   *   this callback is called with no error, it means that the credentials
+   *   information has been loaded into the object (as the `accessKeyId`,
+   *   `secretAccessKey`, and `sessionToken` properties).
+   *   @param err [Error] if an error occurred, this value will be filled
+   * @see get
+   */
+  refresh: function refresh(callback) {
+    this.coalesceRefresh(callback || AWS.util.fn.callback);
+  },
+
+  /**
+   * @api private
+   * @param callback
+   */
+  load: function load(callback) {
+    var self = this;
+    var operation = self.service.config.params.RoleArn ? 'assumeRole' : 'getSessionToken';
+    this.getTokenCode(function (err, tokenCode) {
```

Comment:
Just a thought: supporting MFA will help a lot whenever we use `assumeRole`. This is also what differs from the `TemporaryCredentials`. Generally, I will suggest people using this class over of the `TemporaryCredentials` class.